docroot/modules/humsci/hs_events/src/Services/AutoUnpublish.php

```diff
@@ -0,0 +1,203 @@
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\Core\Datetime\DrupalDateTime;
+use Drupal\Core\Logger\LoggerChannelFactoryInterface;
+
+/**
+ * Service for auto-unpublishing past events and handling form modifications.
+ */
+class AutoUnpublish {
+
+  /**
+   * The entity type manager.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
+
+  /**
+   * The logger factory.
+   *
+   * @var \Drupal\Core\Logger\LoggerChannelFactoryInterface
+   */
+  protected $loggerFactory;
+
+  /**
+   * The config factory.
+   *
+   * @var \Drupal\Core\Config\ConfigFactoryInterface
+   */
+  protected $configFactory;
+
+  /**
+   * Node type for events.
+   */
+  const EVENT_NODE_TYPE = 'hs_event';
+
+  /**
+   * Field name for auto-unpublish setting.
+   */
+  const AUTO_UNPUBLISH_FIELD = 'field_auto_unpublish';
+
+  /**
+   * Field name for event date.
+   */
+  const EVENT_DATE_FIELD = 'field_hs_event_date';
+
+  /**
+   * Config page ID for site options.
+   */
+  const SITE_OPTIONS_CONFIG_ID = 'hs_site_options';
+
+  /**
+   * Field name for site auto-unpublish setting.
+   */
+  const SITE_AUTO_UNPUBLISH_FIELD = 'field_site_auto_unpublish';
+
+  /**
+   * Constructs a new AutoUnpublish service.
+   *
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entity_type_manager
+   *   The entity type manager.
+   * @param \Drupal\Core\Logger\LoggerChannelFactoryInterface $logger_factory
+   *   The logger factory.
+   * @param \Drupal\Core\Config\ConfigFactoryInterface $config_factory
+   *   The config factory.
+   */
+  public function __construct(
+    EntityTypeManagerInterface $entity_type_manager,
+    LoggerChannelFactoryInterface $logger_factory,
+    ConfigFactoryInterface $config_factory,
+  ) {
+    $this->entityTypeManager = $entity_type_manager;
+    $this->loggerFactory = $logger_factory;
+    $this->configFactory = $config_factory;
+  }
+
+  /**
+   * Unpublishes past events that have auto-unpublish enabled.
+   *
+   * @return int
+   *   The number of events that were unpublished.
+   */
+  public function unpublishPastEvents(): int {
+    $node_storage = $this->entityTypeManager->getStorage('node');
+
+    // Get current time in UTC to match database storage.
+    $current_time = new DrupalDateTime('now', 'UTC');
+    $current_timestamp = $current_time->getTimestamp();
+
+    // Query for published hs_event nodes that:
+    // 1. Are published (status = 1).
+    // 2. Have field_auto_unpublish = 1.
+    // 3. Have field_hs_event_date with end time in the past.
+    $query = $node_storage->getQuery()
+      ->condition('type', self::EVENT_NODE_TYPE)
+      ->condition('status', 1)
+      ->condition(self::AUTO_UNPUBLISH_FIELD, 1)
+      ->condition(self::EVENT_DATE_FIELD . '.end_value', $current_timestamp, '<')
+      ->accessCheck(FALSE);
+
+    $nids = $query->execute();
+
+    if (empty($nids)) {
+      return 0;
+    }
+
```

Comment:
Should there be some type of batching here? Either adding a range to the query or only loading a specific item of node id's from storage at a time? I don't think it's absolutely necessary, but it couldn't hurt especially if this gets triggered on a lot of nodes on the first run.